src/Storages/StorageFile.cpp

```diff
@@ -577,29 +716,102 @@ ColumnsDescription StorageFile::getTableStructureFromFile(
         return ColumnsDescription(DistributedAsyncInsertSource(paths[0]).getOutputs().front().getHeader().getNamesAndTypesList());
     }
 
-    if (paths.empty() && !FormatFactory::instance().checkIfFormatHasExternalSchemaReader(format))
+    if (((archive_info && archive_info->paths_to_archives.empty()) || (!archive_info && paths.empty()))
+        && !FormatFactory::instance().checkIfFormatHasExternalSchemaReader(format))
         throw Exception(
             ErrorCodes::CANNOT_EXTRACT_TABLE_STRUCTURE,
             "Cannot extract table structure from {} format file, because there are no files with provided path. "
             "You must specify table structure manually", format);
 
-    std::optional<ColumnsDescription> columns_from_cache;
-    if (context->getSettingsRef().schema_inference_use_cache_for_file)
-        columns_from_cache = tryGetColumnsFromCache(paths, format, format_settings, context);
-
     ColumnsDescription columns;
-    if (columns_from_cache)
+    if (archive_info)
```

Comment:
Maybe we can avoid code duplication a bit here and do smth like:
```c++
std::vector<std::string> paths_for_schema_cache;
std::optional<ColumnsDescription> columns_from_cache;
if (context->getSettingsRef().schema_inference_use_cache_for_file)
{
    if (archive_info)
        columns_from_cache = tryGetColumnsFromCacheForArchives(...);
    else
        columns_from_cache = tryGetColumnsFromCache(...);
}

if (columns_from_cache)
{
    columns = *columns_from_cache;
}
else
{
    if (archive_info)
    {
        ReadBufferFromArchiveIterator read_buffer_iterator(...);
        columns = readSchemaFromFormat(...);
        paths_for_schema_cache = ...;
    }
    else
    {
        ReadBufferFromFilesIterator read_buffer_iterator(...);
        columns = readSchemaFromFormat(...);
        paths_for_schema_cache = ...;
    }
}

 if (context->getSettingsRef().schema_inference_use_cache_for_file)
    addColumnsToCache(paths_for_schema_cache, ...)
```

Or maybe I'm just too perfectionist